blockchyp/blockchyp_client.py

```diff
@@ -101,49 +108,85 @@ def is_terminal_routed(self, request):
     def __gateway_get(self, path):
         url = self.gateway_host + "/api" + path
         gateway_config = self.__get_gateway_config()
-        return requests.get(url, gateway_config)
+        reply = requests.get(url, headers=gateway_config)
+
+        return reply
 
     def __get_gateway_config(self):
-        config = {}
-        if (self.credentials and self.credentials.api_key):
-            # TODO implement CryptoUtils
-            # headers = CryptoUtils.generate_gateway_headers(self.credentials)
-            headers = {
-                "nonce": "",
-                "timestamp": "",
-                "auth_header": ""
-                } # temporary until CryptoUtils is completed
-            config["Headers"] = {
+        if (self.credentials and self.credentials["api_key"]):
+            crypto = CryptoUtils()
+            headers = crypto.generate_gateway_headers(self.credentials)
+            return {
                 "Nonce": headers["nonce"],
                 "Timestamp": headers["timestamp"],
-                "Authorization": headers["auth_header"]
+                "Authorization": headers["auth_header"],
                 }
 
-        return config
-
-    def __get_terminal_config(self):
-        pass
+        return {}
 
     def __gateway_post(self, path, payload):
         pass
 
     def __terminal_get(self, terminal, path, creds):
         pass
 
-    def __terminal_post(self, route, path, payload):
-        pass
+    def __terminal_post(self, route, path, payload=None):
+        url = self.__assemble_terminal_url(route, path)
+        transient_credentials = route["transientCredentials"]
+        data = {
+            "apiKey": transient_credentials["apiKey"],
+            "bearerToken": transient_credentials["bearerToken"],
+            "signingKey": transient_credentials["signingKey"],
+            "request": payload
+        }
+        print(data)
+        reply = requests.post(url, json=data)
+
+        return reply
 
     def __assemble_terminal_url(self, route, path):
-        pass
+        result = "http"
+        if self.https:
+            result = result + "s"
+        result = result + "://"
+        result = result + route["ipAddress"]
+        if self.https:
+            result = result + ":8443"
+        else:
+            result = result + ":8080"
+        result = result + "/api"
+        result = result + path
+
+        return result
 
     def __resolve_terminal_route(self, terminal_name):
-        pass
+        cache_entry = self._route_cache.get(terminal_name, None)
+
+        if cache_entry:
+            # check cache expiration
+            if cache_entry.ttl > int(datetime.utcnow().timestamp()) * 1000:
```

Comment:
Datetime has methods for comparing timestamps with offsets. Don't cast it to an int.